File: IA/Web/WIN-PROLOG_4900/SERVER/INCLUDES/FLEX/Flex.cpp
// Flex.cpp: implementation of the CFlex class.
//
//////////////////////////////////////////////////////////////////////

#include "stdafx.h"
#include "flexeg.h"
#include "Flex.h"
#include "MultBox.h"
#include "LstBox.h"
#include "EdtBox.h"

#ifdef _DEBUG
#undef THIS_FILE
static char THIS_FILE[]=__FILE__;
#define new DEBUG_NEW
#endif

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

// Load flex
// Int386w constructor called first to load prolog
// Then the flex runtime code is loaded through Int386w
const CString errorMessages[] = {
		"isERR_ENCODE",
		"isERR_TICKLE",
		"isERR_CMUTEX",
		"isERR_WMUTEX",
		"isERR_LOCATE",
		"isERR_CREATE",
		"isERR_MAPFIL",
		"isERR_MODULE",
		"isERR_PROLOG",
		"isERR_ACTIVE",
		"isERR_WINDOW" 
};


CFlex::CFlex( LPSTR command, UINT bufferSize, UINT encryption, UINT tickle ) 
   : Int386w( command, bufferSize, encryption, tickle ) {

	if( loadError ) {
		CString errorMessage;

		errorMessage = "Error: " + errorMessages[ abs(loadError) - 1 ] + ", Loading Prolog";

        AfxMessageBox( errorMessage );
		exit( 0 );
	}

	char *resultString = callOneGoal( "ensure_loaded( prolog(flexserv) ). " );

	if( resultString[0] == 'E' ) {
        AfxMessageBox( resultString );
		exit( 0 );
	}
}

CFlex::~CFlex(){

}

char * CFlex::runGoal( char *command ) {
	char *resultString;

	resultString = initGoal( command );

	if( resultString[0] == 'E' ) {
		return( resultString );
		exit(0);
	}

	resultString = callGoal();

	while( resultString[0] == 'I' ) {

		char *textString = &resultString[10];

		switch( resultString[8] ) {
		case 'f' : // Flash a message
			AfxMessageBox( textString );
			break;
		case 'm' : // Multiple selection
			{
				CMltBox dlg;
				dlg.setDialog( textString );
				if ( dlg.DoModal() == IDOK){
					resultString = dlg.m_textString;
				} else {
					resultString = "??. ";
				}

			}
			break;
		case 's' : // Single selection
			{
				CLstBox dlg;
				dlg.setDialog( textString );
				if (dlg.DoModal() == IDOK) {
					resultString = dlg.m_textString;
					strcat( resultString, ". " );
				} else {
					resultString = "??. ";
				}
			}
			break;
		case 'e' : // Edit text
			{
				CEdtBox dlg;
				dlg.setDialog( textString );
				if ( dlg.DoModal() == IDOK) {
					resultString = dlg.m_textString;
				} else {
					resultString = "??. ";
				}
			}
			break;
		case 'x' : // A message box
			char *token;
			char *message;
			int selection, buttons;

			token = strtok( textString, "\001" );
            message  = strtok( NULL, "\001" );
			if( !strcmp( token, "ok" ) )
				buttons = MB_OK;
			else
			    if( !strcmp(token, "okcancel" ) )
				   buttons = MB_OKCANCEL;
			else
				if( !strcmp(token, "yesno") )
					buttons = MB_YESNO;
			else
				if( !strcmp(token, "yesnocancel") )
					buttons = MB_YESNOCANCEL;

            selection = AfxMessageBox(message,buttons);

			switch( selection ) {
			case IDOK:
				resultString = "ok. ";
				break;
			case IDCANCEL:
				resultString = "cancel. ";
				break;
			case IDNO:
				resultString = "no. ";
				break;
			case IDYES:
				resultString = "yes. ";
				break;
			default:
				resultString = "ok. ";
			}

			break;

		default : ;
		}
		resultString = tellGoal( resultString );
	}

	lstrcpy( m_text, resultString );
	exitGoal();

	return( m_text );
}

File: IA/Web/WIN-PROLOG_4900/SERVER/INCLUDES/FLEX/flex.vb
Public Class flex
    Private prolog As IntServer

    Public Sub New(Optional ByVal command As String = "")
        prolog = New IntServer(command, 0, 0, 0)
        prolog.InitGoal("ensure_loaded(prolog(flexserv)). ")
        prolog.CallGoal()
        prolog.ExitGoal()
    End Sub

    Public Function rungoal(ByVal command As String) As String
        'Initialise result string
        Dim result, text, item, rest As String
        Dim buttonid As MsgBoxStyle

        'Initialise the goal
        result = prolog.InitGoal(command)

        'If an error occured return the error
        If result.Substring(0, 1) = "E" Then
            rungoal = result
            Exit Function
        End If

        'Call the initialised valid goal
        result = prolog.CallGoal

        'while flex requires more input
        While result.Substring(0, 1) = "I"
            'get the text
            text = result.Substring(9)
            rest = ""
            ' Select on the input flag
            Select Case result.Substring(8, 1)
                Case "f"    'Flash a message
                    MsgBox(text.Substring(1), MsgBoxStyle.OkOnly)
                    result = ""
                Case "m"    'A multiple selection
                    result = mltbox.multibox(text)
                Case "s"    'A single selection
                    result = lstbox.listbox(text)
                Case "e"  'An edit box
                    result = edtbox.editbox(text)
                Case "x"  'A message box
                    item = flxutil.getitem(text, rest)
                    Select Case item
                        Case "ok"
                            buttonid = MsgBoxStyle.OkOnly
                        Case "okcancel"
                            buttonid = MsgBoxStyle.OkCancel
                        Case "yesno"
                            buttonid = MsgBoxStyle.YesNo
                        Case "yesnocancel"
                            buttonid = MsgBoxStyle.YesNoCancel
                    End Select
                    Select Case MsgBox(rest.Substring(1), buttonid, "Flex Example")
                        Case vbOK
                            result = "ok"
                        Case vbCancel
                            result = "cancel"
                        Case vbYes
                            result = "yes"
                        Case vbNo
                            result = "no"
                    End Select
            End Select
            result = prolog.TellGoal(result + " . ")
        End While
        'Exit the goal
        prolog.ExitGoal()

        'return the result
        rungoal = result
    End Function
    Public Function initGoal(ByVal goal As String) As String
        initGoal = prolog.InitGoal(goal)
    End Function
    Public Function callGoal() As String
        callGoal = prolog.CallGoal()
    End Function
    Public Sub exitgoal()
        prolog.ExitGoal()
    End Sub
End Class
